lightning/src/ln/onchaintx.rs

```diff
@@ -898,45 +895,47 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
+			let holder_commitment_tx = holder_commitment.to_holder_commitment_tx(&self.channel_transaction_parameters, &self.secp_ctx);
+			if let Ok(sigs) = self.key_storage.sign_holder_commitment_htlc_transactions(&holder_commitment_tx, &self.secp_ctx) {
+				// since we are no longer maintaining dust HTLCs info in OnchainTxHandler, all signatures must be present
+				let sigs: Vec<Signature> = sigs.iter().map(|o| o.unwrap()).collect();
 				self.holder_htlc_sigs = Some(Vec::new());
 				let ret = self.holder_htlc_sigs.as_mut().unwrap();
-				for (htlc_idx, (holder_sig, &(ref htlc, _))) in sigs.iter().zip(holder_commitment.per_htlc.iter()).enumerate() {
-					if let Some(tx_idx) = htlc.transaction_output_index {
-						if ret.len() <= tx_idx as usize { ret.resize(tx_idx as usize + 1, None); }
-						ret[tx_idx as usize] = Some((htlc_idx, holder_sig.expect("Did not receive a signature for a non-dust HTLC")));
-					} else {
-						assert!(holder_sig.is_none(), "Received a signature for a dust HTLC");
-					}
-				}
+				Self::populate_holder_htlc_sigs(holder_commitment, sigs, ret)
 			}
 		}
 	}
+
 	fn sign_prev_holder_htlcs(&mut self) {
 		if let Some(ref holder_commitment) = self.prev_holder_commitment {
-			if let Ok(sigs) = self.key_storage.sign_holder_commitment_htlc_transactions(holder_commitment, &self.secp_ctx) {
+			let holder_commitment_tx = holder_commitment.to_holder_commitment_tx(&self.channel_transaction_parameters, &self.secp_ctx);
+			if let Ok(sigs) = self.key_storage.sign_holder_commitment_htlc_transactions(&holder_commitment_tx, &self.secp_ctx) {
+				// since we are no longer maintaining dust HTLCs info in OnchainTxHandler, all signatures must be present
+				let sigs: Vec<Signature> = sigs.iter().map(|o| o.unwrap()).collect();
 				self.prev_holder_htlc_sigs = Some(Vec::new());
 				let ret = self.prev_holder_htlc_sigs.as_mut().unwrap();
-				for (htlc_idx, (holder_sig, &(ref htlc, _))) in sigs.iter().zip(holder_commitment.per_htlc.iter()).enumerate() {
-					if let Some(tx_idx) = htlc.transaction_output_index {
-						if ret.len() <= tx_idx as usize { ret.resize(tx_idx as usize + 1, None); }
-						ret[tx_idx as usize] = Some((htlc_idx, holder_sig.expect("Did not receive a signature for a non-dust HTLC")));
-					} else {
-						assert!(holder_sig.is_none(), "Received a signature for a dust HTLC");
-					}
-				}
+				Self::populate_holder_htlc_sigs(holder_commitment, sigs, ret)
 			}
 		}
 	}
 
+	fn populate_holder_htlc_sigs(holder_commitment: &HolderCommitmentTransaction, sigs: Vec<Signature>, ret: &mut Vec<Option<(usize, Signature)>>) {
+		for (htlc_idx, (holder_sig, htlc)) in sigs.iter().zip(holder_commitment.info.htlcs.iter()).enumerate() {
+			let tx_idx = htlc.transaction_output_index.unwrap();
+			if ret.len() <= tx_idx as usize { ret.resize(tx_idx as usize + 1, None); }
+			ret[tx_idx as usize] = Some((htlc_idx, holder_sig.clone()));
+		}
+	}
```

Comment:
This was existing code that was incidentally DRYed as part of this PR.

`ret` points to of `holder_htlc_sigs` or `prev_holder_htlc_sigs`, and this function mutates them in-place.  If we change it as you suggest, it will not longer be a mutation in-place, so will incur some memory allocation overhead.  Since this is an internal function, it doesn't seem like a big win, but I can change it if you feel that it's confusing.

Reply:
ah, yes. sorry - overuse of IDE powered refactoring on my side.